File: preprocess/qtni_preprocess_config_script.m
%% QTNI for MRI preprocess main script
% v1.3 January 29th 2020
% Mauro Namías - mauro.namias@gmail.com
% This scripts configures and runs the QTNI MRI pre-processing workflow, i.e.:
% spatial normalization to the MNI space, using SPM template.

% v1.4 May 25th 2022
% added step for segment and warp, which bias correct and deformable
% register MRI images

%WITH OR WITHOUT Skull Stripping dependent on which qtni_preprocess file
%you call ("_MRI_wSkullStrip")


%%NOTE: Files must be in .nii format. If in .nii.gz, use gunzip to
%%extract them. Using 7zip or other manual unzippers will change the name
%%of the file.
%Navigate to the directory and type "gunzip -r ." (with the period)

%Notes:
    %%Only add to path folders that you actually want to use. Make sure you
    %don't have any old versions of qnti in the path.
    %%Make sure the temp folder is empty when you start the config script

cfolder = 'C:\Users\deatsch\Documents\MATLAB\QTNI_MRI_preprocess_v1.5\';
path(path,cfolder)
path(path,[cfolder '\Functions\'])



% Configure options
options.temp_path = 'C:\Users\deatsch\Documents\MATLAB\QTNI_MRI_preprocess_v1.5\temp\'  % Directory to store temporal files (which are deleted once the script is finished)
options.TPM_path = 'C:\Users\deatsch\Documents\MATLAB\QTNI_MRI_preprocess_v1.5\Templates\TPM.nii'; % Tissue probability map for brain segmentation and spatial normalization
options.out_path = 'C:\Users\deatsch\Documents\MATLAB\qtni_preprocessed\';  % Output directory, where normalized files are stored
options.PET_template = 'C:\Users\deatsch\Documents\MATLAB\QTNI_MRI_preprocess_v1.5\Templates\single_subj_T1.nii';  % FDG (or any other) brain template for non-linear spatial normaliaztion
options.PET_atlas = 'C:\Users\deatsch\Documents\MATLAB\QTNI_MRI_preprocess_v1.5\Templates\AAL3+Pons\AAL3+pons.nii';
options.intensity_norm_method = 'GM'; % Select either 'GM' (global mean), 'RMS', 'AAL' or 'NONE'. The AAL option allows to select reference regions for normalization (see example in the next line)
% options.intensity_norm_regions = [95:120 171]; % list of AAL regions used for intensity normalization. In this example, the cerebellum and the pons are used. See the \Templates\AAL+pons\AAL3.txt file for the list of region codes
options.SmoothFWHM = 0;  % FWHM of the 3D Gaussian kernel used for smoothing
options.use_oldnorm = 0;  % Use SPM5 spatial normalization (affine + DCT) instead of SPM12 spatial norm (segmentation + warping).


input_dir = 'C:\Users\deatsch\Documents\MATLAB\qtni_input';  % Directory containing all original nifti files

files = dir(input_dir);

progressbar()

    for i = 3:length(files) 
    tic
    input_file = [input_dir '\' files(i).name];
    output_file = qtni_preprocess_MRI_wSkullStrip_noTemplate(input_file, options);    
    progressbar(i/(length(files) ));
    toc
    end



   

File: preprocess/qtni_preprocess_config_script.py
#--------------------------------------------------------------------------------------------------------------#
#                                     Script to preprocess .nii files                                          #
#                                                                                                              #
#                                         Črt Rozman, October 2024                                             #
#                                                                                                              #
#                      Rewritten from Matlab code that was written by Mauro Namies in 2020                     #
#--------------------------------------------------------------------------------------------------------------#



# Importing needed modules
import os
import time
from glob import glob
from tqdm import tqdm  # For progress bar


# Configure options as a dictionary
options = {
    'temp_path': r'C:\Users\deatsch\Documents\MATLAB\QTNI_MRI_preprocess_v1.5\temp\\',  # Temporary file storage
    'TPM_path': r'C:\Users\deatsch\Documents\MATLAB\QTNI_MRI_preprocess_v1.5\Templates\TPM.nii',  # Tissue probability map path
    'out_path': r'C:\Users\deatsch\Documents\MATLAB\qtni_preprocessed\\',  # Output directory
    'PET_template': r'C:\Users\deatsch\Documents\MATLAB\QTNI_MRI_preprocess_v1.5\Templates\single_subj_T1.nii',  # Brain template for normalization
    'PET_atlas': r'C:\Users\deatsch\Documents\MATLAB\QTNI_MRI_preprocess_v1.5\Templates\AAL3+Pons\AAL3+pons.nii',  # Atlas for normalization
    'intensity_norm_method': 'GM',  # Intensity normalization method
    # 'intensity_norm_regions': list(range(95, 121)) + [171],  # Uncomment if needed
    'SmoothFWHM': 0,  # Smoothing parameter
    'use_oldnorm': 0  # Use SPM5 or SPM12 spatial normalization
}


# Directory containing original nifti files
input_dir = r'C:\Users\deatsch\Documents\MATLAB\qtni_input'

# Function placeholder for qtni_preprocess_MRI_wSkullStrip_noTemplate
# You will need to implement this function according to your MRI preprocessing logic
def qtni_preprocess_MRI_wSkullStrip_noTemplate(input_file, options):
    # Preprocessing logic would go here. The function will process the MRI data
    # based on the input file and options, and return an output file.
    # Implement the required steps like segmentation, skull stripping, etc.
    output_file = os.path.join(options['out_path'], os.path.basename(input_file))  # Output path example
    # You can integrate real processing logic here.
    return output_file

# Read all files in the input directory
files = glob(os.path.join(input_dir, '*.nii'))  # Assuming input files are in NIfTI format

# Progress bar for file processing
for i in tqdm(range(len(files)), desc="Processing files"):
    start_time = time.time()
    input_file = files[i]  # Get the file path
    output_file = qtni_preprocess_MRI_wSkullStrip_noTemplate(input_file, options)  # Process the file
    end_time = time.time()
    elapsed_time = end_time - start_time  # Calculate time for the iteration
    print(f"Processed {os.path.basename(input_file)} in {elapsed_time:.2f} seconds")

print("All files processed successfully.")